packages/docusaurus-mdx-loader/src/remark/unusedDirectives/index.ts

```diff
@@ -0,0 +1,72 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+import path from 'path';
+import process from 'process';
+import visit from 'unist-util-visit';
+import logger from '@docusaurus/logger';
+import {posixPath} from '@docusaurus/utils';
+// @ts-expect-error: TODO see https://github.com/microsoft/TypeScript/issues/49721
+import type {Transformer, Processor, Parent} from 'unified';
+import type {
+  Directive,
+  // @ts-expect-error: TODO see https://github.com/microsoft/TypeScript/issues/49721
+} from 'mdast-util-directive';
+
+// TODO as of April 2023, no way to import/re-export this ESM type easily :/
+// This might change soon, likely after TS 5.2
+// See https://github.com/microsoft/TypeScript/issues/49721#issuecomment-1517839391
+// import type {Plugin} from 'unified';
+type Plugin = any; // TODO fix this asap
+
+const directiveTypes = ['containerDirective', 'leafDirective', 'textDirective'];
+
+const plugin: Plugin = function plugin(this: Processor): Transformer {
+  return (tree, file) => {
+    const unusedDirectives: Array<{
+      name: string;
+      type: string;
+      position:
+        | {
+            line: number;
+            column: number;
+          }
+        | undefined;
+    }> = [];
+
+    visit<Parent>(tree, directiveTypes, (node: Directive) => {
+      if (!node.data) {
+        unusedDirectives.push({
+          name: node.name,
+          type: node.type,
+          position: node.position
+            ? {
+                line: node.position.start.line,
+                column: node.position.start.column,
+              }
+            : undefined,
+        });
+      }
+    });
+
+    if (unusedDirectives.length > 0) {
+      const warningMessage = unusedDirectives
+        .map((unusedDirective) => {
+          const positionMessage = unusedDirective.position
+            ? logger.interpolate`number=${unusedDirective.position.line}:number=${unusedDirective.position.column}`
```

Comment:
Note, I'm not sure if IDEs can provide navigation if the line numbers are surrounded by ANSI sequences.

Reply:
For now we agreed navigation was not a goal because we need to use relative paths anyway otherwise CI tests would fail due to different absolute paths in snapshots.

And navigation apparently doesn't work in VSCode with relative paths.

Reply:
I've tested on vscode and in fact there is no problem opening the file from terminal even if the path is relative or surrounded by ASNI sequences (colors in the terminal am I right?)

Reply:
Yes, VS Code supports opening files with paths relative to the current project. I think it even supports arbitrary paths, as long as the name is unambiguous (otherwise it shows a dropdown chooser). For example `foo.tsx` can open to `src/components/foo.tsx`.